### ISSUE TITLE
Falcon Rotator only de-rotates in one direction

### ISSUE THREAD

**Issue by lpldave** [indilib/indi#1933]
When using the derotation feature for the Pegasus Falcon Rotator, it only moves in the counterclockwise direction regardless of whether the direction is set to reversed or not.  

To Reproduce:

Launch the Falcon Rotator driver through the device manager in KStars.  Set value for derotation (I used 10 ms per step).  It rotates in the counterclockwise direction.  Set value for derotation back to zero to disable it.  Click "Enable" under "Reverse" and set value for derotation (I used 10 ms per step).  Rotation is still in the same direction, but is expected to be reversed.

Desktop:

 - OS: PopOS 22.04 LTS (based on Ubuntu)
 - INDILib 2.0.3 and KStars 

Log File:

INFO	7.563281 sec	: Session log file ***
INFO	12.550966 sec	: Pegasus Falcon is online.
INFO	478.595166 sec	: De-rotation is enabled and set to 1 step per 10 milliseconds.
INFO	490.119361 sec	: De-rotaiton is disabled.
INFO	493.575854 sec	: Rotator direction is reversed.
INFO	498.519098 sec	: De-rotation is enabled and set to 1 step per 10 milliseconds.
INFO	523.759419 sec	: De-rotaiton is disabled.
INFO	525.263378 sec	: Rotator direction is normal.
INFO	530.227448 sec	: De-rotation is enabled and set to 1 step per 10 milliseconds.
INFO	555.030551 sec	: De-rotaiton is disabled.
INFO	568.352810 sec	: Pegasus Falcon is offline.

**Comment by escribana** [comment]
Hi @lpldave

I just checked the reaction of my Falcon Rotator respective to the reverse direction and did not notice any problems. Following your suggestion I set the derotation to 10ms per step and changed the rotator direction according your proceeding. The device is reacting as expected, rotates backward if the direction is reversed and forward if set to normal.

OS: Official Ubuntu 22.04 LTS
INDILib 2.0.3 and KStars 3.6.7beta

**Comment by lpldave** [comment]
Thank you for checking!  I’ll give mine another try and see if I can get any more diagnostic info about what might be going on with my setup.

**Comment by escribana** [comment]
Please set the following in "Logging" (button [Logs] in EKOS):

1. Verbosity: Verbose
2. Output: File
3. Drivers: Rotator

Open [Pegasus Falcon]/[Options] in "INDI Control Panel" and set the following:

1. Debug Levels: All (i.e. add "Driver Debug")
2. Logging Levels:  All (i.e. add "Driver Debug")
3. Log Output: "To Log File"

Now the emitted low level commands will be saved to something like this:
[INFO] Session log file /home/<user>/.indi/logs/2023-09-23/indi_falcon_rotator/indi_falcon_rotator_07:38:42.log

**Comment by lpldave** [comment]
Hi, sorry for the huge delay on this.  I tried it again with logging, and the file is attached below.  I tried both reversing in the middle of rotation, as well as stopping rotation, reversing, and restarting.  In all cases, it still only moved in the counterclockwise direction.

INFO	10.833675 sec	: Session log file /home/obrien/.indi/logs/2023-10-27/indi_falcon_rotator/indi_falcon_rotator_08:11:57.log
DEBUG	11.037204 sec	: CMD <FA>
DEBUG	11.076262 sec	: RES <FR_OK:7280:84.03:0:1:0:0
>
DEBUG	12.081354 sec	: CMD <FA>
DEBUG	12.116055 sec	: RES <FR_OK:7280:84.03:0:1:0:0
>
DEBUG	13.121240 sec	: CMD <FA>
DEBUG	13.156261 sec	: RES <FR_OK:7280:84.03:0:1:0:0
>
DEBUG	14.161204 sec	: CMD <FA>
DEBUG	14.197016 sec	: RES <FR_OK:7280:84.03:0:1:0:0
>
DEBUG	15.201199 sec	: CMD <FA>
DEBUG	15.236746 sec	: RES <FR_OK:7280:84.03:0:1:0:0
>
DEBUG	16.241196 sec	: CMD <FA>
DEBUG	16.276979 sec	: RES <FR_OK:7280:84.03:0:1:0:0
>
DEBUG	17.281193 sec	: CMD <FA>
DEBUG	17.317570 sec	: RES <FR_OK:7280:84.03:0:1:0:0
>
DEBUG	18.322811 sec	: CMD <FA>
DEBUG	18.357489 sec	: RES <FR_OK:7280:84.03:0:1:0:0
>
DEBUG	18.805417 sec	: CMD <DR:10>
INFO	18.805485 sec	: De-rotation is enabled and set to 1 step per 10 milliseconds.
DEBUG	19.360483 sec	: CMD <FA>
DEBUG	19.397779 sec	: RES <FR_OK:7304:84.30:1:1:1:0
>
DEBUG	20.401200 sec	: CMD <FA>
DEBUG	20.437990 sec	: RES <FR_OK:7349:84.82:0:1:1:0
>
DEBUG	21.439211 sec	: CMD <FA>
DEBUG	21.478221 sec	: RES <FR_OK:7394:85.34:1:1:1:0
>
DEBUG	22.479639 sec	: CMD <FA>
DEBUG	22.518768 sec	: RES <FR_OK:7439:85.86:1:1:1:0
>
DEBUG	23.520536 sec	: CMD <FA>
DEBUG	23.558740 sec	: RES <FR_OK:7485:86.39:0:1:1:0
>
DEBUG	24.560486 sec	: CMD <FA>
DEBUG	24.599423 sec	: RES <FR_OK:7530:86.91:1:1:1:0
>
DEBUG	25.600699 sec	: CMD <FA>
DEBUG	25.639758 sec	: RES <FR_OK:7575:87.43:1:1:1:0
>
DEBUG	26.641495 sec	: CMD <FA>
DEBUG	26.680008 sec	: RES <FR_OK:7621:87.96:0:1:1:0
>
DEBUG	27.681144 sec	: CMD <FA>
DEBUG	27.720259 sec	: RES <FR_OK:7666:88.48:1:1:1:0
>
DEBUG	28.721686 sec	: CMD <FA>
DEBUG	28.760582 sec	: RES <FR_OK:7711:89.00:1:1:1:0
>
DEBUG	29.762060 sec	: CMD <FA>
DEBUG	29.800478 sec	: RES <FR_OK:7756:89.52:1:1:1:0
>
DEBUG	30.801782 sec	: CMD <FA>
DEBUG	30.841061 sec	: RES <FR_OK:7802:90.05:0:1:1:0
>
DEBUG	31.842844 sec	: CMD <FA>
DEBUG	31.880825 sec	: RES <FR_OK:7847:90.57:1:1:1:0
>
DEBUG	32.882365 sec	: CMD <FA>
DEBUG	32.921553 sec	: RES <FR_OK:7892:91.09:1:1:1:0
>
DEBUG	33.922825 sec	: CMD <FA>
DEBUG	33.961616 sec	: RES <FR_OK:7937:91.61:1:1:1:0
>
DEBUG	34.962844 sec	: CMD <FA>
DEBUG	35.001461 sec	: RES <FR_OK:7983:92.14:1:1:1:0
>
DEBUG	36.002684 sec	: CMD <FA>
DEBUG	36.041728 sec	: RES <FR_OK:8028:92.66:1:1:1:0
>
DEBUG	36.461474 sec	: CMD <FN:1>
DEBUG	36.489857 sec	: RES <FN:1
>
INFO	36.489937 sec	: Rotator direction is reversed.
DEBUG	37.042970 sec	: CMD <FA>
DEBUG	37.082308 sec	: RES <FR_OK:8024:92.61:0:1:1:1
>
DEBUG	38.084079 sec	: CMD <FA>
DEBUG	38.122898 sec	: RES <FR_OK:7979:92.09:1:1:1:1
>
DEBUG	39.124667 sec	: CMD <FA>
DEBUG	39.162683 sec	: RES <FR_OK:7934:91.57:1:1:1:1
>
DEBUG	40.163978 sec	: CMD <FA>
DEBUG	40.202762 sec	: RES <FR_OK:7888:91.04:0:1:1:1
>
DEBUG	41.204020 sec	: CMD <FA>
DEBUG	41.243498 sec	: RES <FR_OK:7843:90.52:1:1:1:1
>
DEBUG	42.245238 sec	: CMD <FA>
DEBUG	42.283697 sec	: RES <FR_OK:7798:90.00:0:1:1:1
>
DEBUG	43.285466 sec	: CMD <FA>
DEBUG	43.323962 sec	: RES <FR_OK:7753:89.49:1:1:1:1
>
DEBUG	44.325652 sec	: CMD <FA>
DEBUG	44.364125 sec	: RES <FR_OK:7707:88.95:0:1:1:1
>
DEBUG	45.365697 sec	: CMD <FA>
DEBUG	45.403926 sec	: RES <FR_OK:7662:88.43:0:0:1:1
>
DEBUG	46.405189 sec	: CMD <FA>
DEBUG	46.444456 sec	: RES <FR_OK:7617:87.92:1:0:1:1
>
DEBUG	47.446194 sec	: CMD <FA>
DEBUG	47.484923 sec	: RES <FR_OK:7572:87.40:1:0:1:1
>
DEBUG	48.486671 sec	: CMD <FA>
DEBUG	48.525188 sec	: RES <FR_OK:7527:86.88:1:0:1:1
>
DEBUG	49.526458 sec	: CMD <FA>
DEBUG	49.564849 sec	: RES <FR_OK:7481:86.35:0:0:1:1
>
DEBUG	50.566606 sec	: CMD <FA>
DEBUG	50.605463 sec	: RES <FR_OK:7436:85.83:1:0:1:1
>
DEBUG	51.417406 sec	: CMD <FN:0>
DEBUG	51.437637 sec	: RES <FN:0
>
INFO	51.437751 sec	: Rotator direction is normal.
DEBUG	51.605916 sec	: CMD <FA>
DEBUG	51.645789 sec	: RES <FR_OK:7406:85.48:0:0:1:0
>
DEBUG	52.647136 sec	: CMD <FA>
DEBUG	52.685841 sec	: RES <FR_OK:7451:86.00:1:0:1:0
>
DEBUG	53.687577 sec	: CMD <FA>
DEBUG	53.726432 sec	: RES <FR_OK:7497:86.53:1:0:1:0
>
DEBUG	54.728213 sec	: CMD <FA>
DEBUG	54.766770 sec	: RES <FR_OK:7542:87.05:1:0:1:0
>
DEBUG	55.768577 sec	: CMD <FA>
DEBUG	55.806397 sec	: RES <FR_OK:7587:87.57:1:0:1:0
>
DEBUG	56.807695 sec	: CMD <FA>
DEBUG	56.847164 sec	: RES <FR_OK:7633:88.10:1:0:1:0
>
DEBUG	57.848939 sec	: CMD <FA>
DEBUG	57.887459 sec	: RES <FR_OK:7679:88.63:0:0:1:0
>
DEBUG	58.889237 sec	: CMD <FA>
DEBUG	58.927276 sec	: RES <FR_OK:7724:89.15:1:0:1:0
>
DEBUG	59.928542 sec	: CMD <FA>
DEBUG	59.967342 sec	: RES <FR_OK:7769:89.67:0:0:1:0
>
DEBUG	60.968534 sec	: CMD <FA>
DEBUG	61.007566 sec	: RES <FR_OK:7814:90.19:1:0:1:0
>
DEBUG	62.008759 sec	: CMD <FA>
DEBUG	62.047820 sec	: RES <FR_OK:7860:90.72:0:0:1:0
>
DEBUG	63.049491 sec	: CMD <FA>
DEBUG	63.088754 sec	: RES <FR_OK:7905:91.24:0:0:1:0
>
DEBUG	64.089177 sec	: CMD <FA>
DEBUG	64.128310 sec	: RES <FR_OK:7950:91.76:0:0:1:0
>
DEBUG	64.128633 sec	: CMD <DR:0>
INFO	64.128712 sec	: De-rotaiton is disabled.
DEBUG	65.129180 sec	: CMD <FA>
DEBUG	65.168602 sec	: RES <FR_OK:7952:91.78:0:0:0:0
>
DEBUG	66.169183 sec	: CMD <FA>
DEBUG	66.208808 sec	: RES <FR_OK:7952:91.78:0:0:0:0
>
DEBUG	67.209187 sec	: CMD <FA>
DEBUG	67.249063 sec	: RES <FR_OK:7952:91.78:0:0:0:0
>
DEBUG	68.045412 sec	: CMD <FN:1>
DEBUG	68.065393 sec	: RES <FN:1
>
INFO	68.065435 sec	: Rotator direction is reversed.
DEBUG	68.249405 sec	: CMD <FA>
DEBUG	68.289281 sec	: RES <FR_OK:7952:91.78:0:0:0:1
>
DEBUG	69.289468 sec	: CMD <FA>
DEBUG	69.329572 sec	: RES <FR_OK:7952:91.78:0:0:0:1
>
DEBUG	70.330765 sec	: CMD <FA>
DEBUG	70.369854 sec	: RES <FR_OK:7952:91.78:0:0:0:1
>
DEBUG	71.371539 sec	: CMD <FA>
DEBUG	71.410378 sec	: RES <FR_OK:7952:91.78:0:0:0:1
>
DEBUG	72.411710 sec	: CMD <FA>
DEBUG	72.450415 sec	: RES <FR_OK:7952:91.78:0:0:0:1
>
DEBUG	72.949498 sec	: CMD <DR:10>
INFO	72.949595 sec	: De-rotation is enabled and set to 1 step per 10 milliseconds.
DEBUG	73.451190 sec	: CMD <FA>
DEBUG	73.491145 sec	: RES <FR_OK:7930:91.53:1:0:1:1
>
DEBUG	74.492898 sec	: CMD <FA>
DEBUG	74.531450 sec	: RES <FR_OK:7885:91.01:1:0:1:1
>
DEBUG	75.533217 sec	: CMD <FA>
DEBUG	75.571648 sec	: RES <FR_OK:7839:90.48:0:0:1:1
>
DEBUG	76.573407 sec	: CMD <FA>
DEBUG	76.611591 sec	: RES <FR_OK:7794:89.96:1:0:1:1
>
DEBUG	77.613143 sec	: CMD <FA>
DEBUG	77.652130 sec	: RES <FR_OK:7749:89.44:1:0:1:1
>
DEBUG	78.653868 sec	: CMD <FA>
DEBUG	78.692080 sec	: RES <FR_OK:7703:88.91:0:0:1:1
>
DEBUG	79.693386 sec	: CMD <FA>
DEBUG	79.732544 sec	: RES <FR_OK:7658:88.39:0:0:1:1
>
DEBUG	80.733836 sec	: CMD <FA>
DEBUG	80.772815 sec	: RES <FR_OK:7613:87.87:1:0:1:1
>
DEBUG	81.774582 sec	: CMD <FA>
DEBUG	81.812808 sec	: RES <FR_OK:7568:87.35:0:0:1:1
>
DEBUG	82.814322 sec	: CMD <FA>
DEBUG	82.853306 sec	: RES <FR_OK:7523:86.83:1:0:1:1
>
DEBUG	83.855051 sec	: CMD <FA>
DEBUG	83.893627 sec	: RES <FR_OK:7477:86.30:1:0:1:1
>
DEBUG	84.895384 sec	: CMD <FA>
DEBUG	84.933758 sec	: RES <FR_OK:7432:85.78:1:0:1:1
>
DEBUG	85.935522 sec	: CMD <FA>
DEBUG	85.973496 sec	: RES <FR_OK:7387:85.26:1:0:1:1
>
DEBUG	86.974753 sec	: CMD <FA>
DEBUG	87.013627 sec	: RES <FR_OK:7341:84.73:1:0:1:1
>
DEBUG	88.014911 sec	: CMD <FA>
DEBUG	88.053883 sec	: RES <FR_OK:7296:84.21:0:0:1:1
>
DEBUG	89.055152 sec	: CMD <FA>
DEBUG	89.094091 sec	: RES <FR_OK:7251:83.69:0:0:1:1
>
DEBUG	90.095290 sec	: CMD <FA>
DEBUG	90.134324 sec	: RES <FR_OK:7206:83.17:1:0:1:1
>
DEBUG	91.135518 sec	: CMD <FA>
DEBUG	91.174882 sec	: RES <FR_OK:7161:82.65:1:0:1:1
>
DEBUG	92.176090 sec	: CMD <FA>
DEBUG	92.214882 sec	: RES <FR_OK:7115:82.12:1:0:1:1
>
DEBUG	93.037428 sec	: CMD <DR:0>
INFO	93.037520 sec	: De-rotaiton is disabled.
DEBUG	93.215242 sec	: CMD <FA>
DEBUG	93.255142 sec	: RES <FR_OK:7077:81.68:0:0:0:1
>
DEBUG	94.256329 sec	: CMD <FA>
DEBUG	94.295406 sec	: RES <FR_OK:7077:81.68:0:0:0:1
>
DEBUG	94.701430 sec	: CMD <FN:0>
DEBUG	94.727479 sec	: RES <FN:0
>
INFO	94.727521 sec	: Rotator direction is normal.
DEBUG	95.296147 sec	: CMD <FA>
DEBUG	95.335594 sec	: RES <FR_OK:7077:81.68:0:0:0:0
>
DEBUG	96.336765 sec	: CMD <FA>
DEBUG	96.375870 sec	: RES <FR_OK:7077:81.68:0:0:0:0
>
DEBUG	97.377005 sec	: CMD <FA>
DEBUG	97.416152 sec	: RES <FR_OK:7077:81.68:0:0:0:0
>
DEBUG	98.417617 sec	: CMD <FA>
DEBUG	98.456455 sec	: RES <FR_OK:7077:81.68:0:0:0:0
>
DEBUG	99.457207 sec	: CMD <FA>
DEBUG	99.496654 sec	: RES <FR_OK:7077:81.68:0:0:0:0
>

**Comment by escribana** [comment]
Hi @lpldave

The analysis of the log file shows that all is correct!
Look here:  [RS-LogAnalysis.txt](https://github.com/indilib/indi/files/13197597/RS-LogAnalysis.txt)
Thus I don't know why you experience a wrong behaviour while the device is reporting a correct one.  I'm out of ideas!

**Comment by lpldave** [comment]
Thanks for checking.  I haven’t seen that particular error report you linked, but I have seen a few others.  Looks like it’s basically the same issue I’m having.  I’ll try testing this on a Windows machine when I get a chance, and send a message to tech support.

**Comment by lpldave** [comment]
I was able to get the Windows driver running in a virtual machine under PopOS/Ubuntu.  The derotation worked in both directions, so it doesn't seem to be a hardware problem.

**Comment by escribana** [comment]
Ok, so it seems to be a publication problem!!! (-> [Forum](https://pegasusastro.com/forum/viewtopic.php?t=91).
The published  [Falcon_Serial_Command_Table.pdf](https://github.com/indilib/indi/files/13400359/Falcon_Serial_Command_Table.pdf) seems not to be complete. I don't know why they want to hide certain commands.:-(
The only way to reveal these would be a very time consuming reverse engeneering.

**Comment by lpldave** [comment]
I figured out what the problem was.  The driver only takes positive values of the de-rotation parameter (milliseconds per step), which drives the rotator in the counterclockwise direction.  Negative values are required to move it in the clockwise direction.  I confirmed this by directly sending commands to it over the USB port.  A command of "DR:1" as I used above gives 1 ms/step rotation in the counterclockwise direction.   A command of "DR:-1" gives 1 ms/step rotation in the clockwise direction.

So the driver should be modified to allow for both positive and negative values of the de-rotation parameter.
